src/file_system.rs

```diff
@@ -233,31 +190,96 @@ fn metadata() {
+                fs::read_link(&path_buf)?
+            }
+        };
+        path_buf.pop();
+        if fs::symlink_metadata(&path_buf)?.is_symlink() {
+            path_buf = fast_canonicalize(path_buf.as_path())?;
+        }
+        for component in link.components() {
+            match component {
+                Component::ParentDir => {
+                    path_buf.pop();
+                }
+                Component::Normal(seg) => {
+                    #[cfg(target_family = "wasm")]
+                    {
+                        // Need to trim the extra \0 introduces by https://github.com/nodejs/uvwasi/issues/262
+                        path_buf.push(seg.to_string_lossy().trim_end_matches('\0'));
                     }
+                    #[cfg(not(target_family = "wasm"))]
+                    {
+                        path_buf.push(seg);
+                    }
+                }
+                Component::RootDir => {
+                    #[cfg(not(windows))]
+                    {
+                        path_buf = PathBuf::from("/");
+                    }
+                    #[cfg(windows)]
+                    {
+                        path_buf.push("");
+                    }
+                }
+                Component::Prefix(prefix) => {
+                    path_buf = PathBuf::from(prefix.as_os_str());
                 }
-                _ => Ok(canonicalized),
+                Component::CurDir => {}
+            }
+
+            if fs::symlink_metadata(&path_buf)?.is_symlink() {
+                path_buf = fast_canonicalize(path_buf.as_path())?;
             }
-        } else {
-            fs::canonicalize(path)
         }
+        if !fs::symlink_metadata(&path_buf)?.is_symlink() {
+            break;
+        }
+    }
+    Ok(path_buf)
+}
+
+#[cfg(windows)]
+fn node_compatible_raw_canonicalize<P: AsRef<Path>>(path: P) -> PathBuf {
+    use std::path::{Component, Prefix};
+    let canonicalized = path
+        .as_ref()
+        .as_os_str()
+        .as_encoded_bytes()
+        .strip_prefix(UNC_PATH_PREFIX).map_or_else(|| path.as_ref().to_path_buf(), |p| {
+            // SAFETY: `as_encoded_bytes` ensures `p` is valid path bytes
+            unsafe { PathBuf::from(std::ffi::OsStr::from_encoded_bytes_unchecked(p)) }
+        });
+    let first_component = canonicalized.components().next();
+    match first_component {
+        Some(Component::Prefix(prefix)) => match prefix.kind() {
+            Prefix::VerbatimUNC(_, _) => canonicalized
+                .to_str()
+                .and_then(|s| s.get(UNC_PATH_PREFIX.len()..))
+                .map(PathBuf::from)
+                .unwrap_or(canonicalized),
+            Prefix::VerbatimDisk(_) => canonicalized
+                .to_str()
+                .and_then(|s| s.get(LONG_PATH_PREFIX.len()..))
+                .map(PathBuf::from)
+                .unwrap_or(canonicalized),
+            _ => canonicalized,
+        },
+        _ => canonicalized,
```

Comment:
I guess we should use `strip_prefix` for `LONG_PATH_PREFIX` as well and remove this part.